### PR TITLE
Feature: Fork jobs

### PR DESCRIPTION
**Pull request by kratob**
This PR is a bit on the larger side, but can probably be reviewed easily enough commit by commit.
If need be, I could split it into several smaller PRs, but I'd rather save the effort, if not necessary.

Requirements:

> Problem:
> Errands run in thread can still fuck up the whole scheduler
> 
> Solution:
> - Add a :fork => :thread | :process option to the runner DSL
> - Make it default (keep the threading variant)
> - When forking a process, and AR is present, reconnect after the fork
> - Forked processes to report success/failure through their exit status (0 = success)
> - Logging content unchanged, but add file locking around logging + flushing (to avoid partial/cluttered log lines)
> - Timeout management handled in the master process, which should send SIGKILL on timeout
> 
> Locking:
> - create a lockfile /tmp/rufus-<master pid>.lock
> - lock it (File.flock) whenever preparing to write a log line
> - unlock it afterwards

### CODE REVIEW

**Comment by kratob** [comment]
(Closing, since not tested on staging, yet.)

**Comment by mezis** [comment]
File locking does not work across threads, only processes.

**Comment by mezis** [comment]
> File locking does not work across threads, only processes

Which means you need a mutex around your file lock :)

**Comment by kratob** [comment]
Seems to work fine on my machine, even for threads:

``` ruby
Thread.new do
  f = File.new("/tmp/foo", "w")
  f.flock(File::LOCK_EX)
  puts "inner lock acquired"
  sleep 5
  puts "inner lock released"
  f.flock(File::LOCK_UN)
end

f = File.new("/tmp/foo", "w")
sleep 1
f.flock(File::LOCK_EX)
puts "outer lock acquired"

# > inner lock acquired
# > inner lock released
# > outer lock acquired
```

**Comment by kratob** [comment]
Works on staging.

**Comment by mezis** [comment]
Nice big cleanup while you were at it :)
Test coverage is good, and it does work for me too ; so once you're fed up nagging people for a review, merge away ^_^

File: lib/rufus-runner/locking_io_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end
+
+
+  def initialize(*)
+    @guard = Guard.new
+    super
+  end
+
+  %w(write puts print).each do |method|
```
Comment (pedrocunha) [inline lib/rufus-runner/locking_io_decorator.rb]: fancy way of declaring simple methods?

File: lib/rufus-runner/tracking_scheduler.rb [hunk 2]
```diff
@@ -17,38 +22,28 @@ def initialize(options = {})
...
     return unless rails_environment_matches?(options.delete(:environments))
-
-    name  = options.delete(:name) || 'noname'
-    if frequency = options.delete(:every)
-      scheduling_method = :every
-    elsif frequency = options.delete(:cron)
-      scheduling_method = :cron
+    options = options.merge(@options)
+
+    name = options.delete(:name) || 'noname'
```
Comment (pedrocunha) [inline lib/rufus-runner/tracking_scheduler.rb:27]: honestly I don't like this code pattern. You are deletting keys from the object that is given as argument. The caller might have a little surprise after he runs this method.

File: lib/rufus-runner/tracking_scheduler.rb [hunk 3]
```diff
@@ -128,6 +146,7 @@ def rails_environment
   end
 
   DefaultOptions = {
+    :fork              => :process,  # safety first
```
Comment (pedrocunha) [inline lib/rufus-runner/tracking_scheduler.rb:149]: "safety first" - I laughed.

File: lib/rufus-runner/tracking_scheduler/forking_job_runner.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+class Rufus::TrackingScheduler::ForkingJobRunner < Rufus::TrackingScheduler::JobRunner
+
+  class UnexpectedExitStatus < StandardError; end
+
+  def run_block
+    @pid = fork do
+      @block.call
+    end
+    status = Process.wait2(@pid)[1]
```
Comment (pedrocunha) [inline lib/rufus-runner/tracking_scheduler/forking_job_runner.rb:11]: wait2. Nice method name

File: lib/rufus-runner/tracking_scheduler/job_runner.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  private
+
+  def run_block
+    raise NotImplementedError.new
+  end
+
+  def job_id
+    @job_id ||= '%08x' % @job.job_id.gsub(/\D/,'')
```
Comment (pedrocunha) [inline lib/rufus-runner/tracking_scheduler/job_runner.rb:49]: Some explanation here on this char would be nice :D

File: spec/locking_io_decorator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    let(:lock_file) { double('lock_file', :flock => true) }
+
+    before do
+      subject.stub :lock_file => lock_file
+    end
+
+    it 'locks and unlocks a lock file' do
+      subject.stub(:lock_file => lock_file)
+      reached = false
+      lock_file.should_receive(:flock).with(File::LOCK_EX)
```
Comment (pedrocunha) [inline spec/locking_io_decorator_spec.rb]: Isn't this an implementation detail?

File: spec/locking_io_decorator_spec.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      subject.sync do
+        reached = true
+        lock_file.should_receive(:flock).with(File::LOCK_UN)
+      end
+      reached.should be_true
+    end
+
+    it 'releases locks on errors' do
+      lock_file.should_receive(:flock).with(File::LOCK_UN)
```
Comment (pedrocunha) [inline spec/locking_io_decorator_spec.rb]: same comment as before

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -78,6 +83,45 @@ def expect_new_file(pathname)
...
+  def process_running?(pid)
+    status = Process.getpgid(pid)
+    true
+  rescue Errno::ESRCH
+    false
+  end
+
+  def wait_for_child_processes
+    loop do
+      Process.wait(-1)
```
Comment (pedrocunha) [inline spec/spec_helper.rb]: What does -1 do on wait? Doesn't wait at all? But leaves the CPU for another task?

**Comment by pedrocunha** [comment]
This looks nice. For me, it's a shame some parts of code aren't more documented to fully understand what's going on. Had some few comments but overall looks very good and solid.

File: lib/rufus-runner/tracking_scheduler.rb (hunk 2 above)
Comment (kratob) [inline lib/rufus-runner/tracking_scheduler.rb:27]: `options` is actually duped (due to the `merge`), so a caller will not notice. I cannot just keep the params in there either, or rufus-scheduler will complain about "invalid arguments" later.

File: lib/rufus-runner/tracking_scheduler/forking_job_runner.rb (hunk 4 above)
Comment (kratob) [inline lib/rufus-runner/tracking_scheduler/forking_job_runner.rb:11]: The whole `Process` class is a train wreck.

File: lib/rufus-runner/tracking_scheduler/job_runner.rb (hunk 5 above)
Comment (kratob) [inline lib/rufus-runner/tracking_scheduler/job_runner.rb:49]: Indeed. Problem is, I have no clue either (this code is old, was just moved around). :laughing:

File: lib/rufus-runner/locking_io_decorator.rb (hunk 1 above)
Comment (kratob) [inline lib/rufus-runner/locking_io_decorator.rb]: Shall change that.

File: spec/locking_io_decorator_spec.rb (hunk 6 above)
Comment (kratob) [inline spec/locking_io_decorator_spec.rb]: You're right, and I don't particularly like it, either.
But the only alternative that does not test implementation details is to spawn a number of processes, let them have a go at writing to stdout, and checking that output is not garbled.
Do you think that would be better?

File: spec/spec_helper.rb (hunk 8 above)
Comment (kratob) [inline spec/spec_helper.rb]: "Random child". I'll add a comment.

File: lib/rufus-runner/tracking_scheduler.rb (hunk 2 above)
Comment (pedrocunha) [inline lib/rufus-runner/tracking_scheduler.rb:27]: Ok :+1:

File: lib/rufus-runner/tracking_scheduler/forking_job_runner.rb (hunk 4 above)
Comment (pedrocunha) [inline lib/rufus-runner/tracking_scheduler/forking_job_runner.rb:11]: Thought so

File: lib/rufus-runner/tracking_scheduler/job_runner.rb (hunk 5 above)
Comment (pedrocunha) [inline lib/rufus-runner/tracking_scheduler/job_runner.rb:49]: hahah fair enough :D

File: spec/locking_io_decorator_spec.rb (hunk 6 above)
Comment (pedrocunha) [inline spec/locking_io_decorator_spec.rb]: Maybe not. But if that flock blah blah was inside a more descriptive method name, the should receive probably would be more readable? Just a suggestion.

Comment (kratob) [inline spec/locking_io_decorator_spec.rb]: But at the end of the day, I need to either test the effect of the `flock` (which is hard), or treat it like an external API and test that it is called. If I wrap it in a descriptive method, I'd have to spec that method, too.
I shall add a comment to improve readability.

Comment (pedrocunha) [inline spec/locking_io_decorator_spec.rb]: Fair enough. Documentation should suffice.

File: lib/rufus-runner/tracking_scheduler/forking_job_runner.rb (hunk 4 above)
Comment (mezis) [inline lib/rufus-runner/tracking_scheduler/forking_job_runner.rb:11]: Hum `wait2` is actually [part of POSIX](http://linux.die.net/man/2/wait). So Ruby does a good job of clearly mapping to the standard. Now, some have been known to call POSIX a train wreck ...

**Pull request merged**